### PR TITLE
WalletConnect no internet edge-cases

### LINKED ISSUE #20760: WalletConnect requests should fail if not connected to internet
WalletConnect requests should show a "not connected" error toast when the user tries to sign/send the request.

cc: @xAlisher can we have an extra case for toasts when there is no internet?

**Steps:** 

1. Connect to dApp
2. Trigger some event
3. On mobile device enable Airplane mode
4. Sign the event 
5. Observe the result

**Actual result:** UI does not reflect the fact user is signing transaction while not being connected to the internet

https://github.com/user-attachments/assets/d745295f-1951-44eb-977d-20deb5cbda3c

**Expected result:** we should inform user that he is not connected to internet and therefore transaction/message cannot be signed.

_Originally posted by @pavloburykh in https://github.com/status-im/status-mobile/issues/20693#issuecomment-2228710185_

### LINKED ISSUE #20764: Initialize the WalletConnect SDK if failed during login due to internet connection issues
# Bug Report

## Problem 
[comment]: # (A problem description)
If a user has no internet connection during login, the sdk will not be initialized and the user won't be able to use the wallet connect functionality. We should detect if the internet connection appears and re-initialize wallet connect.

#### Expected behavior
Wallet connect works if logged in without internet but got internet afterwards

#### Actual behavior
Wallet doesn't work if logged in without internet but got internet afterwards

### Reproduction

1. Turn off wifi and mobile data
2. Log in
3. Turn internet back on
4. Try to connect or receive session requests (should fail)

### Additional Information

- Status version: nightly 
- Operating System: Android, iOS

[comment]: # (Please, add logs/notes if necessary)

### LINKED ISSUE #20802: No error is shown when connecting to dApp without internet connection
Originally reported here https://github.com/status-im/status-mobile/pull/20711#issuecomment-2236244746

**Steps:** 

1. Enable airplane mode on mobile device
2. Scan wallet connect QR and tap connect
3. See if any error is shown informing that connection has not been established

**Expected result:** we should inform user that connection has not been established because network is unavailable

**Actual result:** no error

https://github.com/user-attachments/assets/872bcd40-4696-46e1-bdcf-577338e006c9

### Additional Information

- Status version: nightly 
- Operating System: Android, iOS

### PR DESCRIPTION
**Pull request by clauxx**
fixes #20802 
fixes #20760 
fixes #20764

### Summary

Fixing issues related to internet connection in wallet connect.

**Changes included:**
1. Moved the `net-info` ns from legacy to status-im and adapted to guidelines
2. Moved network (internet) root subs from legacy to status-im
3. If the user logs in without internet, show the persisted sessions in "Connected dApps"
4. If the user logs in without internet, don't initialise wallet-connect until the connection is re-established
5. Show toast if disconnecting without internet connection
6. Show toast if approving dapp connection without internet
7. Show toast if scanning wallet-connect QR without internet
8. Show alert banner inside request modal (above slider, similar to the fee warning) when connection is lost. Not using a toast here, since the request can only come in with an internet connection, so we're just handling the case where the connection is lost while the modal is open. The slider is disabled while there is no connection.
9. For point 8, moved the quo/alert-banner into the reusable footer component

status: ready

### CODE REVIEW

File: src/status_im/contexts/wallet/wallet_connect/events.cljs [hunk 1]
```diff
@@ -282,3 +309,12 @@
...
             :on-success #(log/info "Wallet Connect session disconnected")
             :on-error   #(log/info "Wallet Connect session persistence failed" %)}]]]}))
+
+(rf/reg-event-fx
+ :wallet-connect/no-internet-toast
+ (fn []
+   {:fx [[:dispatch
+          [:toasts/upsert
+           {:type  :negative
+            :theme :dark
```
Comment (J-Son89) [inline src/status_im/contexts/wallet/wallet_connect/events.cljs]: is it always theme dark? 🤔

File: src/status_im/contexts/wallet/wallet_connect/modals/common/footer/view.cljs [hunk 2]
```diff
@@ -13,23 +13,32 @@
...
-     [rn/view {:style style/warning-container}
-      [quo/text
-       {:size   :paragraph-2
-        :style  {:color colors/neutral-80-opa-70}
-        :weight :medium}
-       warning-label]]]))
+  [{:keys [warning-label slide-button-text error-text]} & children]
+  (let [{:keys [customization-color]} (rf/sub [:wallet-connect/current-request-account-details])
+        network-status                (rf/sub [:network/status])
+        offline?                      (= network-status :offline)]
```
Comment (J-Son89) [inline src/status_im/contexts/wallet/wallet_connect/modals/common/footer/view.cljs]: I thought we had some agreement to make subs for cases like these where it does the accessor to prevent rerenders when the other data changes.
i.e the sub should return offline or not, or likely it needs a new sub with just that info.

**Review by J-Son89 — APPROVED** [review]

File: src/status_im/contexts/wallet/wallet_connect/events.cljs (hunk 1 above)
Comment (clauxx) [reply src/status_im/contexts/wallet/wallet_connect/events.cljs]: uff, good point 👍

File: src/status_im/contexts/wallet/wallet_connect/modals/common/footer/view.cljs (hunk 2 above)
Comment (clauxx) [reply src/status_im/contexts/wallet/wallet_connect/modals/common/footer/view.cljs]: you're right. This would apply to both subs here it seems. For the first one we may need a sub specific for the color. As for the network status, I remember we shouldn't use root subs inside components. I'll make a layer 2 sub for it that returns a boolean 👍. Thank you!

File: src/status_im/contexts/wallet/wallet_connect/events.cljs (hunk 1 above)
Comment (clauxx) [reply src/status_im/contexts/wallet/wallet_connect/events.cljs]: done

File: src/status_im/contexts/wallet/wallet_connect/modals/common/footer/view.cljs (hunk 2 above)
Comment (clauxx) [reply src/status_im/contexts/wallet/wallet_connect/modals/common/footer/view.cljs]: done

**Review by alwx — APPROVED** [review]

**Comment by clauxx** [comment]
@qoqobolo @pavloburykh The PR is ready for testing

**Comment by qoqobolo** [comment]
@clauxx could you resolve the conflicts please? I'll start testing after that, thanks!

**Comment by clauxx** [comment]
@qoqobolo done 👍

**Comment by ilmotta** [comment]
Hey @clauxx, I found quite hard to review this PR due to the amount of commits coming from other already merged PRs. Even the large number of changes coming from the recently changed translations file.

Sorry if I'm out of context, but why are we basing the PR branch off of `fix/typed-data-wrong-chain` and not `develop`? From a QA perspective this also makes things complicated/risky to test.

**Comment by clauxx** [comment]
> Hey @clauxx, I found quite hard to review this PR due to the amount of commits coming from other already merged PRs. Even the large number of changes coming from the recently changed translations file.
> 
> 
> 
> Sorry if I'm out of context, but why are we basing the PR branch off of `fix/typed-data-wrong-chain` and not `develop`? From a QA perspective this also makes things complicated/risky to test.

Yeah, good point. I have to rebase. Initially the changes here depended on the other branch

**Comment by clauxx** [comment]
@ilmotta Sorry for this, seems like during a previous rebase the tree got messed up. Should be good to review and test now 👍

**Comment by qoqobolo** [comment]
@clauxx `tests` are failing again :( check them please

**Comment by clauxx** [comment]
> @clauxx `tests` are failing again :( check them please

seems to be the linting issue again. Should be fixed now

**Comment by qoqobolo** [comment]
Works great, thanks @clauxx!

Just FYI: there is still one edge case when you close the Status app with a pending request (e.g. `send transaction`), turn off internet, log in, and turn on the Internet.

As a result, you get zeros on your balance, since we do not trigger an automatic balance update after the internet connection has returned (I'm going to log this feature request separately).

And this in turn leads to the fact that when the WalletConnect modal with a request is opened automatically when re-connecting to internet, an alert is displayed that you do not have enough balance to complete this transaction.

WDYT about this scenario, should we report this and do we want/can handle it somehow?

**Comment by clauxx** [comment]
> Works great, thanks @clauxx!
> 
> Just FYI: there is still one edge case when you close the Status app with a pending request (e.g. `send transaction`), turn off internet, log in, and turn on the Internet.
> 
> As a result, you get zeros on your balance, since we do not trigger an automatic balance update after the internet connection has returned (I'm going to log this feature request separately).
> 
> And this in turn leads to the fact that when the WalletConnect modal with a request is opened automatically when re-connecting to internet, an alert is displayed that you do not have enough balance to complete this transaction.
> 
> WDYT about this scenario, should we report this and do we want/can handle it somehow?

Thanks, will check soon if it's difficult to add.

File: src/status_im/common/device_network.cljs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+(rf/reg-event-fx
+ :device-network/on-state-change
+ (fn [{:keys [db]} [{:keys [isConnected type details]}]]
+   (let [old-network-status       (:network/status db)
+         old-network-type         (:network/type db)
+         connectivity-status      (if isConnected :online :offline)
+         status-changed?          (not= connectivity-status old-network-status)
+         type-changed?            (not= type old-network-type)
+         is-connection-expensive? (:is-connection-expensive details)]
+     (log/debug "[net-info]"
```
Comment (ilmotta) [inline src/status_im/common/device_network.cljs]: I don't know if it's important, but sharing this handler is very similar to the one in https://github.com/status-im/status-mobile/pull/20730/files#diff-efce04a1c18c0acfd8ece4416c70f868204e0a4557c5f649bea69d1c4311380cR44

File: src/status_im/contexts/wallet/wallet_connect/events.cljs [hunk 4]
```diff
@@ -236,16 +261,18 @@
...
-                                      :namespaces
-                                      :eip155
-                                      :accounts)))
-                         sessions)]
-     {:fx [[:dispatch [:wallet-connect/fetch-active-sessions]]]
+   (let [network-status (:network/status db)
+         sessions'      (mapv (fn [{:keys [sessionJson] :as session}]
+                                (assoc session
+                                       :accounts
+                                       (-> sessionJson
```
Comment (ilmotta) [inline src/status_im/contexts/wallet/wallet_connect/events.cljs:264]: Here a bit better to grab accounts using `oget` and only then convert the result to CLJ.

**Comment by clauxx** [comment]
@qoqobolo I can't really test it in debug as I can't resume the connection to debug builds when re-enabling internet. Can you please check if the issue is fixed when the build is ready? If not, I'll remove the commit and we'll make a separate issue for this bug.

File: src/status_im/contexts/wallet/wallet_connect/events.cljs [hunk 5]
```diff
@@ -236,16 +261,18 @@
...
-                                  :accounts
-                                  (-> sessionJson
-                                      types/json->clj
-                                      :namespaces
-                                      :eip155
-                                      :accounts)))
-                         sessions)]
-     {:fx [[:dispatch [:wallet-connect/fetch-active-sessions]]]
+   (let [network-status (:network/status db)
+         sessions'      (mapv (fn [{:keys [sessionJson] :as session}]
```
Comment (ilmotta) [inline src/status_im/contexts/wallet/wallet_connect/events.cljs:261]: Do you really need the eager version of `map`?

File: src/status_im/common/device_network.cljs [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    [native-module.core :as native-module]
+    [status-im.feature-flags :as ff]
+    [taoensso.timbre :as log]
+    [utils.re-frame :as rf]))
+
+(rf/reg-fx
+ :effects.device-network/listen-to-network-info
+ (fn []
+   (when net-info
+     (.addEventListener ^js net-info
```
Comment (ilmotta) [inline src/status_im/common/device_network.cljs]: FYI This event listener logic is also present in this open PR https://github.com/status-im/status-mobile/pull/20730/files#diff-d1c789ab23b32d0ccbb4d5cf3dc0d6ba6fc33cbcc65c33d26ca9bde71956ef48R5

**Review by ilmotta — APPROVED** [review]

File: src/status_im/common/device_network.cljs (hunk 3 above)
Comment (clauxx) [reply src/status_im/common/device_network.cljs]: ahh what are the odds 😆

File: src/status_im/contexts/wallet/wallet_connect/events.cljs (hunk 5 above)
Comment (clauxx) [reply src/status_im/contexts/wallet/wallet_connect/events.cljs:261]: not my addition, but most likely not

**Comment by qoqobolo** [comment]
> @qoqobolo I can't really test it in debug as I can't resume the connection to debug builds when re-enabling internet. Can you please check if the issue is fixed when the build is ready? If not, I'll remove the commit and we'll make a separate issue for this bug.

I checked on https://react-app.walletconnect.com/ on Mainnet, but I'm not sure what's the expected result now?🤔

In the latest build it works like this:
1. Connect to the dapp (user should have some tokens)
2. Trigger `eth_sendTransaction`, and wait for the request modal
3. Close the app with the pending request
4. Turn off internet
5. Log in
6. Turn on internet 

In the previous build, the modal was re-opened at this moment, but now this does not happen. Balances are also not updated.
So it looks like the last commit didn't fix the problem.
But I also don’t know what exactly you intended to do: not show the modal or fix the balance update itself?

I still suggest logging and fixing this case it separately, since this is a pretty rare case, and I'm afraid we may spend too much time on this now, what do you think?

File: src/status_im/common/device_network.cljs (hunk 3 above)
Comment (clauxx) [reply src/status_im/common/device_network.cljs]: should've merged when I had the chance :). This is the same change as there, basically just moved the legacy code to status-im. I'll check with @Parveshdhull.

**Comment by clauxx** [comment]
> > @qoqobolo I can't really test it in debug as I can't resume the connection to debug builds when re-enabling internet. Can you please check if the issue is fixed when the build is ready? If not, I'll remove the commit and we'll make a separate issue for this bug.
> 
> I checked on https://react-app.walletconnect.com/ on Mainnet, but I'm not sure what's the expected result now?🤔
> 
> In the latest build it works like this:
> 
> 1. Connect to the dapp (user should have some tokens)
> 2. Trigger `eth_sendTransaction`, and wait for the request modal
> 3. Close the app with the pending request
> 4. Turn off internet
> 5. Log in
> 6. Turn on internet
> 
> In the previous build, the modal was re-opened at this moment, but now this does not happen. Balances are also not updated. So it looks like the last commit didn't fix the problem. But I also don’t know what exactly you intended to do: not show the modal or fix the balance update itself?
> 
> I still suggest logging and fixing this case it separately, since this is a pretty rare case, and I'm afraid we may spend too much time on this now, what do you think?

Yes, makes sense. I intended to update the balances, but seems like it didn't work. Will remove the commit.

On another note, seems like there's another PR that moves the network functionality from legacy, same as I did. Gotta figure what we do so we don't end up with duplicate code.

**Comment by qoqobolo** [comment]
Thanks @clauxx, ping me please as soon as you have news👍

File: src/status_im/contexts/wallet/wallet_connect/events.cljs (hunk 4 above)
Comment (clauxx) [reply src/status_im/contexts/wallet/wallet_connect/events.cljs:264]: `sessionJson` is stringified JSON, so not sure we can avoid the conversion to JSON here. We could parse the string to JSON and then use `oget`, but do we gain much in perf?

**Comment by clauxx** [comment]
@qoqobolo ready to be checked again. Added some changes to make it easier for the other PR to rebase (cc @Parveshdhull).

**Comment by qoqobolo** [comment]
Thanks @clauxx, LGTM.
PR can be merged.
